crates/lib/src/execution_data.rs

```diff
@@ -63,44 +126,95 @@ impl ReadoutMap {
         readout_mappings: &HashMap<String, String>,
         readout_values: &HashMap<String, ReadoutValues>,
     ) -> Self {
-        let result = readout_values
-            .iter()
-            .flat_map(|(readout_name, values)| {
-                readout_mappings
-                    .iter()
-                    .filter(|(_, program_alias)| *program_alias == readout_name)
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
+        let mut result = ReadoutMap(HashMap::new());
+        for (readout_name, values) in readout_values {
+            readout_mappings
+                .iter()
+                .filter(|(_, program_alias)| *program_alias == readout_name)
+                .map(|(program_name, _)| program_name.as_ref())
+                .map(parse_readout_register)
+                .filter_map(Result::ok)
+                .for_each(|reference| {
+                    let row = match &values.values {
+                        Some(readout_values::Values::IntegerValues(ints)) => ints
+                            .values
+                            .clone()
+                            .into_iter()
+                            .map(|i| Some(ReadoutValue::I32(i)))
+                            .collect(),
+                        Some(readout_values::Values::ComplexValues(comps)) => comps
+                            .values
+                            .clone()
+                            .into_iter()
+                            .map(|c| Complex64::new(c.real().into(), c.imaginary().into()))
+                            .map(|c| Some(ReadoutValue::Complex64(c)))
+                            .collect(),
+                        None => Vec::new(),
+                    };
+                    // TODO handle possible truncation
+                    let shape = (reference.index as usize + 1, row.len());
+                    let matrix = result
+                        .0
+                        .entry(reference.name)
+                        .and_modify(|m| {
+                            if shape.0 > m.nrows() {
```

Comment:
Possible off by one?

Reply:
I think this one is okay, the purpose of `shape` is to describe the number of rows and columns the matrix should have, so on l160 we define the number of rows as `reference.index + 1`. If that number is equal to the existing matrices `nrows`, then there is already enough space to accommodate a row at `reference.index`